### PR TITLE
[GenDeviceCodeLoader] Recursively adding all dependent kernels' Quake code

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

We have a logic to add all dependent kernel quake code but it only works for first-level calls.

This adds a recursive step to chase the call chain if needed.

A sample use case is Python MLIR [uccsd](https://github.com/NVIDIA/cuda-quantum/blob/main/python/cudaq/kernels/uccsd.py ) where there are multi-level call chains, e.g., `uccsd` -> `uccsd_odd_electrons` -> `single_excitation`.

### CODE REVIEW

File: test/NVQPP/gen_device_code.cpp
Comment (bmhowe23) [inline test/NVQPP/gen_device_code.cpp:1]: I think that per the `tests/README.txt` file, this should probably go in the `targettests/execution` directory.

**Review by bmhowe23 — APPROVED** [review]
Nice find!

File: test/NVQPP/gen_device_code.cpp
Comment (1tnguyen) [reply test/NVQPP/gen_device_code.cpp:1]: I've moved the file to the `targettests/execution` directory

File: lib/Optimizer/Transforms/GenDeviceCodeLoader.cpp [hunk 1]
```diff
@@ -91,22 +91,27 @@ class GenerateDeviceCodeLoader
...
-            if (includedNeededFuncs.contains(neededFunc.getName()))
-              return WalkResult::skip();
-            neededFunc.print(strOut, opf);
-            includedNeededFuncs.insert(neededFunc.getName());
-            strOut << '\n';
-            return WalkResult::advance();
-          }
-          return WalkResult::advance();
-        });
+        const std::function<WalkResult(func::CallOp)> emitDependentFunc =
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/GenDeviceCodeLoader.cpp:94]: This variable isn't really needed.

Comment (bmhowe23) [reply lib/Optimizer/Transforms/GenDeviceCodeLoader.cpp:94]: I'm curious ... if this variable isn't needed, how would you perform the recursive call in line 108?

Comment (schweitzpgi) [reply lib/Optimizer/Transforms/GenDeviceCodeLoader.cpp:94]: We're already walking every function in the module. This just turns a linear scan into a recursive dive back through the same functions. Why do that?

Create a map for each function in turn. Then scan that map and create the transitive closure.

Comment (1tnguyen) [reply lib/Optimizer/Transforms/GenDeviceCodeLoader.cpp:94]: Thanks @schweitzpgi. Now I see your point. I added #1366 to get it fixed.

**Pull request merged**